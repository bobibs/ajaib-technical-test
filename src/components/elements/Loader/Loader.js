import React from 'react';
import styles from './styles.module.css';

export default function Loader() {
  return (
    <div className={styles.root}>
      <div></div>
      <div></div>
      <div></div>
    </div>
  );
}
