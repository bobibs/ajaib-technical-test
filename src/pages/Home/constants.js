export const FAILED = 'HOME/FAILED';
export const LOADING = 'HOME/LOADING';
export const SUCCESS = 'HOME/SUCCESS';
