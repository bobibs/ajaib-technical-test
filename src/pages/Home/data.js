import moment from 'moment';
import { capitalizeFirstLetter, textTruncate } from '../../utils/format';

export const filterGenderList = [
  { label: 'Male', value: 'male' },
  { label: 'Female', value: 'female' },
];

export const convertData = (data) => {
  return data.results.map((i) => [
    {
      text: textTruncate(
        `${i.name.title.toLowerCase()}${i.name.first.toLowerCase()}`,
        20,
      ),
      key: 'username',
    },
    { text: textTruncate(`${i.name.first} ${i.name.last}`, 20), key: 'name' },
    { text: textTruncate(i.email, 20), key: 'email' },
    { text: capitalizeFirstLetter(i.gender), key: 'gender' },
    {
      text: moment(i.registered.date).format('DD/MM/YYYY'),
      key: 'registeredDate',
    },
  ]);
};

export const listTableHead = [
  {
    title: 'Username',
    key: 'username',
  },
  {
    title: 'Name',
    key: 'name',
    ascDesc: true,
    handleAsc: () => console.log('asc'),
    handleDesc: () => console.log('desc'),
  },
  {
    title: 'Email',
    key: 'email',
    ascDesc: true,
    handleAsc: () => console.log('asc'),
    handleDesc: () => console.log('desc'),
  },
  {
    title: 'Gender',
    key: 'gender',
    ascDesc: true,
    handleAsc: () => console.log('asc'),
    handleDesc: () => console.log('desc'),
  },
  {
    title: 'Registered Date',
    key: 'registeredDate',
    ascDesc: true,
    handleAsc: () => console.log('asc'),
    handleDesc: () => console.log('desc'),
  },
];
