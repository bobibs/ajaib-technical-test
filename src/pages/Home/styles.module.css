.root {
    height: 100vh;
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
    padding: 5rem;
}

.filter {
    display: flex;
    flex-direction: row;
    margin: 0 0 1rem 0;
}

.filterInput,
.filterSelect {
    margin: 0 1rem 0 0;
}

.filterPagination {
    width: 100%;
    display: flex;
    justify-content: flex-end;
    margin: 2rem 0 0 0;
}