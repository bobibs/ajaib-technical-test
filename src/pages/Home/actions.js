import { FAILED, LOADING, SUCCESS } from './constants';
import { getUsers } from '../../utils/fetch';

export const getAllUsers = (value) => {
  const val = {
    gender: value.gender,
    page: value.page,
    search: value.search,
  };

  return (dispatch) => {
    dispatch(loadingAction(true));
    getUsers(val)
      .then((res) => {
        dispatch(successAction(res.data));
        dispatch(loadingAction(false));
      })
      .catch((err) => {
        dispatch(failedAction(err.message));
        dispatch(loadingAction(false));
      });
  };
};

const failedAction = (isLoading, message) => {
  return { type: FAILED, isLoading, message };
};

const loadingAction = (isLoading) => {
  return { type: LOADING, isLoading };
};
const successAction = (data) => {
  return { type: SUCCESS, data };
};
