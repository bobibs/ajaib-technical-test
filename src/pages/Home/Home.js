import React, { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import moment from 'moment';
import Button from '../../components/elements/Button';
import Loader from '../../components/elements/Loader';
import Table from '../../components/elements/Table';
import InputSearch from '../../components/fields/InputSearch';
import Pagination from '../../components/fields/Pagination';
import SelectOption from '../../components/fields/SelectOption';
import { filterGenderList, listTableHead } from './data';
import styles from './styles.module.css';

export default function Home() {
  const { data, getAllUsers } = useSelector((s) => s.homeReducer);
  const [resetSearch, setResetSearch] = useState(false);
  const [resetGender, setResetGender] = useState(false);
  const [resetPagination, setResetPagination] = useState(false);
  const [filterGender, setFilterGender] = useState('');
  const [filterSearch, setFilterSearch] = useState('');

  const dispatch = useDispatch();

  const resetAllValue = () => {
    setResetSearch(false);
    setResetGender(false);
    setResetPagination(false);

    setTimeout(() => {
      setResetSearch(true);
      setResetGender(true);
      setResetPagination(true);
    }, 500);

    const val = {
      gender: '',
      page: 1,
    };

    dispatch(getAllUsers(val));
  };

  const getValueSearch = (val) => {
    setFilterSearch(val);
    setResetPagination(false);

    const vals = {
      gender: filterGender,
      page: 1,
    };

    dispatch(getAllUsers(vals));
  };

  const getValueGender = (val) => {
    setFilterGender(val);
    setResetPagination(false);
    const vals = {
      gender: val,
      page: 1,
    };

    dispatch(getAllUsers(vals));
  };

  const onClickPagination = (page) => {
    const val = {
      gender: filterGender,
      page,
    };

    dispatch(getAllUsers(val));
  };

  useEffect(() => {
    setResetPagination(true);
  }, [filterGender, filterSearch]);

  useEffect(() => {
    dispatch(getAllUsers(1));
  }, []);

  if (!Object.keys(data).length) {
    return (
      <div className={styles.root}>
        <Loader />
      </div>
    );
  }

  return (
    <div className={styles.root}>
      <div className={styles.filter}>
        <InputSearch
          className={styles.filterInput}
          onClick={getValueSearch}
          resetValue={resetSearch}
        />
        <SelectOption
          className={styles.filterSelect}
          onClick={getValueGender}
          optionList={filterGenderList}
          resetValue={resetGender}
        />
        <Button size='small' variant='outlined' onClick={resetAllValue}>
          Reset Filter
        </Button>
      </div>
      <Table column={listTableHead} dataSource={data} />
      <div className={styles.filterPagination}>
        <Pagination
          onClick={onClickPagination}
          resetValue={resetPagination}
          totalData={data.results ? data.results.length : 10}
        />
      </div>
    </div>
  );
}
