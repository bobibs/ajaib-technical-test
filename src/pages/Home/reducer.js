import { FAILED, LOADING, SUCCESS } from './constants';
import { getAllUsers } from './actions';

const initialState = {
  data: {},
  getAllUsers,
  isLoading: false,
  message: '',
};

export default function reducer(state = initialState, action = {}) {
  const { data, isLoading, message, type } = action;

  switch (type) {
    case FAILED:
      return {
        ...state,
        message,
      };

    case LOADING:
      return {
        ...state,
        isLoading,
      };

    case SUCCESS:
      return {
        ...state,
        data,
      };

    default:
      return state;
  }
}
