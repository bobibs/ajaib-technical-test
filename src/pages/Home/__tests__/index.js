import React, { useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import ShallowRenderer from 'react-test-renderer/shallow';
import Home from '../Home';
import Loader from '../../../components/elements/Loader';

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: jest.fn((fn) => fn()),
  useSelector: jest.fn((fn) => fn()),
}));

jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useState: jest.fn(),
  useEffect: jest.fn(),
}));

const dispatch = jest.fn();

const setResetSearch = jest.fn();
const setResetGender = jest.fn();
const setResetPagination = jest.fn();
const setFilterGender = jest.fn();
const setFilterSearch = jest.fn();

describe('src/components/elements/Home', () => {
  beforeEach(() => {
    useState.mockImplementation(() => [false, setResetSearch]);
    useState.mockImplementation(() => [false, setResetGender]);
    useState.mockImplementation(() => [false, setResetPagination]);
    useState.mockImplementation(() => ['', setFilterGender]);
    useState.mockImplementation(() => ['', setFilterSearch]);
    useDispatch.mockImplementation(() => dispatch);
    useSelector.mockImplementation((fn) => {
      fn({});
      return {
        data: {
          results: [
            {
              name: 'name',
            },
          ],
        },
        getAllUsers: jest.fn(),
      };
    });
  });

  test('render', () => {
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<Home />);
    expect(tree).toMatchSnapshot();
  });

  test('loading', () => {
    useSelector.mockImplementationOnce(() => ({
      data: {},
    }));

    const result = Home();
    expect(result.props.children.type).toBe(Loader);
  });

  test('function getValueSearch', () => {
    const result = Home();
    result.props.children[0].props.children[0].props.onClick();
    expect(dispatch).toHaveBeenCalled();
  });

  test('function getValueGender', () => {
    const result = Home();
    result.props.children[0].props.children[1].props.onClick();
    expect(dispatch).toHaveBeenCalled();
  });

  test('function getValueGender', () => {
    const result = Home();
    result.props.children[0].props.children[2].props.onClick();
    expect(dispatch).toHaveBeenCalled();
  });

  test('function clickPagination', () => {
    const result = Home();
    result.props.children[2].props.children.props.onClick();
    expect(dispatch).toHaveBeenCalled();
  });
});
