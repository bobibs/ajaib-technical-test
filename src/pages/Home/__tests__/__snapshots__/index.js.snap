// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/components/elements/Home render 1`] = `
<div
  className="root"
>
  <div
    className="filter"
  >
    <InputSearch
      className="filterInput"
      onClick={[Function]}
      placeholder="Search"
      resetValue=""
    />
    <SelectOption
      className="filterSelect"
      onClick={[Function]}
      optionList={
        Array [
          Object {
            "label": "Male",
            "value": "male",
          },
          Object {
            "label": "Female",
            "value": "female",
          },
        ]
      }
      resetValue=""
    />
    <Button
      className=""
      onClick={[Function]}
      size="small"
      variant="outlined"
    >
      Reset Filter
    </Button>
  </div>
  <Table
    column={
      Array [
        Object {
          "key": "username",
          "title": "Username",
        },
        Object {
          "ascDesc": true,
          "handleAsc": [Function],
          "handleDesc": [Function],
          "key": "name",
          "title": "Name",
        },
        Object {
          "ascDesc": true,
          "handleAsc": [Function],
          "handleDesc": [Function],
          "key": "email",
          "title": "Email",
        },
        Object {
          "ascDesc": true,
          "handleAsc": [Function],
          "handleDesc": [Function],
          "key": "gender",
          "title": "Gender",
        },
        Object {
          "ascDesc": true,
          "handleAsc": [Function],
          "handleDesc": [Function],
          "key": "registeredDate",
          "title": "Registered Date",
        },
      ]
    }
    dataSource={
      Object {
        "results": Array [
          Object {
            "name": "name",
          },
        ],
      }
    }
  />
  <div
    className="filterPagination"
  >
    <Pagination
      className=""
      onClick={[Function]}
      resetValue=""
      totalData={1}
    />
  </div>
</div>
`;
